modules/http/src/main/java/org/mule/module/http/internal/listener/grizzly/ResponseStreamingCompletionHandler.java

```diff
@@ -66,28 +66,37 @@ public void sendInputStreamChunk() throws IOException
     {
         final Buffer buffer = memoryManager.allocate(DEFAULT_BUFFER_SIZE);
 
-        final byte[] bufferByteArray = buffer.array();
         final int offset = buffer.arrayOffset();
         final int length = buffer.remaining();
 
-        int bytesRead = inputStream.read(bufferByteArray, offset, length);
-        final HttpContent content;
-
-        if (bytesRead == -1)
-        {
-            content = httpResponsePacket.httpTrailerBuilder().build();
-            isDone = true;
-        }
-        else
-        {
-            buffer.limit(bytesRead);
-            content = httpResponsePacket.httpContentBuilder().content(buffer).build();
-        }
+        isDone = readStreamManually(buffer);
 
+        HttpContent content = httpResponsePacket.httpContentBuilder().content(buffer).build();
         ctx.getConnection().getAttributes().setAttribute(MULE_CLASSLOADER, loggerClassLoader);
         ctx.write(content, this);
+
+        if (isDone) {
+            content = httpResponsePacket.httpContentBuilder().build();
+            ctx.write(content, this);
+        }
+    }
+
+    private boolean readStreamManually(final Buffer buffer) throws IOException {
+        buffer.clear();
+        boolean isDone = false;
+        byte[] bufferByteArray = buffer.array();
```

Comment:
Im always forcing this to write from index 0, I have done a clear() and then printed arrayOffset. It was set to 0

Reply:
Sorry but I have to disagree.
The `clear` method does not necessary affect the `arrayOffset`. The clear method will clear the position and other flags of course, but for most `Buffer` implementations, it can't change the `arrayOffset`. Remember that the underlying array (if it exists), may be shared by multiple `Buffer` views.

What exactly happens will, of course, depend on the actual implementation that you are getting. But we should not write our code assuming any specific implementation here.

The test you made is incomplete, because you are printing the `arrayOffset` **after** clearing and it was 0, but you never printed it **before** clearing it, it might very well have been 0 already.

Here you can see an example in which the array offset is not cleared after calling the `clear` method: https://ideone.com/XGgUoC

The example uses a `ByteArray` and a slice of it.